### PR TITLE
Bugfix/5165 enable pl optimizer

### LINKED ISSUE #5159: performance loss from 1.0.8 to 1.1.* when using 16 bit precision
## 🐛 Bug

After updating pytorch-lightning from 1.0.8 to 1.1.0/1.1.1 the use of 16 bit precision destroys the performances.
In my actual code of object detection losses are by a factor of 4 larger at the beginning than compared to 32 bit or 16 bit with pl 1.08.
They converge to a much higher value and the resulting model lost its detection capabilities completely.
To replicate I tested the pl notebooks and the `06-cifar10-baseline.ipynb` also shows this and the classification accuracy corresponds to guessing the class when switching from 32 to 16 bit.
I integrated it into the BoringModel notebook and the problem is also happening in google colab.

## Please reproduce using [the BoringModel and post here](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing)

https://colab.research.google.com/drive/1FqXG9Xw9gVZxnwiGnjsHpAtb-vUqFaob?usp=sharing

### Expected behavior

Same performance for 32 and 16 bit.

### Environment

* CUDA:
	- GPU:
		- Tesla P100-PCIE-16GB
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.18.5
	- pyTorch_debug:     True
	- pyTorch_version:   1.7.0+cu101
	- pytorch-lightning: 1.1.1
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.6.9
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020

### LINKED ISSUE #5165: Reproducibility bug between LightningOptimizer activate / deactivate
## Please reproduce using [the BoringModel and post here](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing)

```
import os
import torch
from torch.utils.data import Dataset
from pytorch_lightning import Trainer, LightningModule, seed_everything


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):

    def __init__(self):
        """
        Testing PL Module

        Use as follows:
        - subclass
        - modify the behavior for what you want

        class TestModel(BaseTestModel):
            def training_step(...):
                # do your own thing

        or:

        model = BaseTestModel()
        model.training_epoch_end = None

        """
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def loss(self, batch, prediction):
        # An arbitrary loss to have a loss that updates the model weights during `Trainer.fit` calls
        return torch.nn.functional.mse_loss(prediction, torch.ones_like(prediction))

    def step(self, x):
        x = self.layer(x)
        out = torch.nn.functional.mse_loss(x, torch.ones_like(x))
        return out

    def training_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        print(loss)
        return {"loss": loss}

    def training_step_end(self, training_step_outputs):
        return training_step_outputs

    def training_epoch_end(self, outputs) -> None:
        torch.stack([x["loss"] for x in outputs]).mean()

    def validation_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"x": loss}

    def validation_epoch_end(self, outputs) -> None:
        torch.stack([x['x'] for x in outputs]).mean()

    def test_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"y": loss}

    def test_epoch_end(self, outputs) -> None:
        torch.stack([x["y"] for x in outputs]).mean()

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
        return [optimizer], [lr_scheduler]


#  NOTE: If you are using a cmd line to run your script,
#  provide the cmd line as below.
#  opt = "--max_epochs 1 --limit_train_batches 1".split(" ")
#  parser = ArgumentParser()
#  args = parser.parse_args(opt)

def run_test():
    seed_everything(42)

    # fake data
    train_data = torch.utils.data.DataLoader(RandomDataset(32, 64))
    val_data = torch.utils.data.DataLoader(RandomDataset(32, 64))

    # model
    before = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        max_epochs=1,
        limit_train_batches=4,
        limit_val_batches=0,
        weights_summary=None,
        gpus=1,
        precision=16,
    )
    trainer.fit(before, train_data, val_data)

    seed_everything(42)
    # fake data
    train_data = torch.utils.data.DataLoader(RandomDataset(32, 64))
    val_data = torch.utils.data.DataLoader(RandomDataset(32, 64))

    # model
    after = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        max_epochs=1,
        limit_train_batches=4,
        limit_val_batches=0,
        weights_summary=None,
        gpus=1,
        precision=16,
        enable_pl_optimizer=False
    )
    trainer.fit(after, train_data, val_data)

    # Assert model parameters are identical after fit
    for before, after in zip(before.parameters(), after.parameters()):
        assert torch.equal(before, after), 'Model parameters are different'


if __name__ == '__main__':
    run_test()
```
Credit to @SeanNaren 

### Environment

**Note**: `Bugs with code` are solved faster ! `Colab Notebook` should be made `public` !

* `IDE`: Please, use our python [bug_report_model.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py
) template.

* `Colab Notebook`: Please copy and paste the output from our [environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py) (or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
# For security purposes, please check the contents of collect_env_details.py before running it.
python collect_env_details.py
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #5165
Fixes #5159

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/core/test_lightning_optimizer.py [hunk 1]
```diff
@@ -433,3 +433,109 @@ def configure_optimizers(self):
...
+        trainer = Trainer(
+            default_root_dir=tmpdir,
+            max_epochs=max_epochs,
+            limit_train_batches=limit_train_batches,
+            limit_val_batches=limit_val_batches,
+            enable_pl_optimizer=enable_pl_optimizer,
+            accumulate_grad_batches=accumulate_grad_batches if not override else 1
+        )
+        trainer.fit(model)
+        assert trainer.global_step == expected_global_step
```
Comment (SeanNaren) [inline tests/core/test_lightning_optimizer.py:507]: I don't think this assert is needed in this test? I guess it doesn't hurt but this test now tests a couple of additional things already.

File: tests/core/test_lightning_optimizer.py [hunk 2]
```diff
@@ -433,3 +433,109 @@ def configure_optimizers(self):
...
+                    if batch_nb % accumulate_grad_batches == 0:
+                        optimizer.step(closure)
+
+            model = TestModel("SGD")
+        else:
+            expected_global_step = (max_epochs * limit_train_batches) // accumulate_grad_batches
+            if enable_pl_optimizer:
+                model = TestBoringModel("Adam" if mock else "SGD")
+            else:
+                model = TestBoringModel("AdamW" if mock else "SGD")
```
Comment (SeanNaren) [inline tests/core/test_lightning_optimizer.py:497]: Not sure what's going on here, why does the PL optimizer control if it's adam/adamw?

Comment (tchaton) [reply tests/core/test_lightning_optimizer.py:497]: So I can use mock and assert exactly the call_count

File: tests/core/test_lightning_optimizer.py (hunk 1 above)
Comment (tchaton) [reply tests/core/test_lightning_optimizer.py:507]: Just to make sure it trained as expected. global_step is wrong in the automatic_optimization case with optimizer_step overrides.

File: pytorch_lightning/plugins/native_amp.py [hunk 3]
```diff
@@ -52,7 +53,10 @@ def backward(self, closure_loss, optimizer, opt_idx, *args, **kwargs):
 
         # unscale gradient to allow analyze within `on_after_backward`
         if not self.trainer.train_loop.should_accumulate() and automatic_optimization:
-            self.trainer.scaler.unscale_(optimizer)
+            if isinstance(optimizer, LightningOptimizer):
+                self.trainer.scaler.unscale_(optimizer._optimizer)
```
Comment (Borda) [inline pytorch_lightning/plugins/native_amp.py:57]: shall we rather have define in PL opt `def __call__() -> optimizer._optimizer`?

File: tests/core/test_lightning_optimizer.py [hunk 4]
```diff
@@ -433,3 +435,175 @@ def configure_optimizers(self):
...
+    initial_weights = {}
+
+    class TestBoringModel(BoringModel):
+
+        def __init__(self, optimizer_name):
+            super().__init__()
+            self._optimizer_name = optimizer_name
+            self.losses = []
+            self.grads = []
+            self.on_before_zero_grad_count = 0
```
Comment (Borda) [inline tests/core/test_lightning_optimizer.py:460]: would it be better to mock it?
https://medium.com/python-pandemonium/python-mocking-you-are-a-tricksy-beast-6c4a1f8d19b2
but maybe not needed now...

File: tests/core/test_lightning_optimizer.py [hunk 5]
```diff
@@ -433,3 +435,175 @@ def configure_optimizers(self):
...
+    assert override.losses == before.losses
+    assert (override.on_before_zero_grad_count // accumulate_grad_batches) == before.on_before_zero_grad_count
+
+    for b_grad, o_grad in zip(before.grads, override.grads):
+        assert torch.equal(b_grad, o_grad), 'Grad parameters are different'
+
+    for b_w, o_w in zip(before.parameters(), override.parameters()):
+        assert torch.equal(b_w, o_w), 'Model parameters are different'
+
+    if precision == 32:
```
Comment (Borda) [inline tests/core/test_lightning_optimizer.py:591]: is there a version for precision 16

File: tests/core/test_lightning_optimizer.py [hunk 6]
```diff
@@ -433,3 +435,175 @@ def configure_optimizers(self):
...
+    max_epochs = np.random.randint(1, 3)
+    limit_train_batches = np.random.randint(11, 27)
+
+    print(max_epochs, limit_train_batches)
+
+    expected_batches = max_epochs * limit_train_batches
+    is_divisible = limit_train_batches % accumulate_grad_batches == 0
+    limit_val_batches = 0
+
+    def train(enable_pl_optimizer, override=False, mock=False):
```
Comment (Borda) [inline tests/core/test_lightning_optimizer.py:488]: this seems to very long, can we rather cut this function out?

File: tests/core/test_lightning_optimizer.py [hunk 7]
```diff
@@ -433,3 +435,175 @@ def configure_optimizers(self):
...
+])
+@pytest.mark.parametrize('accumulate_grad_batches', [1])
+def test_parity_training_lightning_optimizer(tmpdir, amp_backend, precision, accumulate_grad_batches):
+    """
+    Test training with accumulated gradients with and within enable_pl_optimizer reaches the same weights
+    """
+
+    initial_weights = {}
+
+    class TestBoringModel(BoringModel):
```
Comment (Borda) [inline tests/core/test_lightning_optimizer.py:453]: can we also cut this class out, the test is very heavy now...

File: pytorch_lightning/trainer/training_loop.py [hunk 8]
```diff
@@ -501,6 +501,9 @@ def optimizer_step(self, optimizer, opt_idx, batch_idx, train_step_and_backward_
             using_lbfgs=is_lbfgs,
         )
 
+        if not self.trainer._enable_pl_optimizer and is_overridden("optimizer_step", model_ref):
+            self.on_before_zero_grad(optimizer)
```
Comment (SeanNaren) [inline pytorch_lightning/trainer/training_loop.py:505]: Maybe add a note here that this happens automatically within the lightning optimizer. Now thinking of it, what about:

```
if self.require_manual_optimizer_hook_call:
    self.on_before_zero_grad(optimizer)
```

**Comment by SeanNaren** [comment]
@tchaton could you also just revert this commit on your branch:

https://github.com/PyTorchLightning/pytorch-lightning/commit/3910b283399f1b7888eecc63ab88ac375c284752

File: pytorch_lightning/plugins/native_amp.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/plugins/native_amp.py]: I move the _unscale to optimizer_step. Let's see if it doesn't break the test.

File: tests/core/test_lightning_optimizer.py (hunk 4 above)
Comment (tchaton) [reply tests/core/test_lightning_optimizer.py:462]: Could do, but I am also capture the gradients to make sure everything is properly designed.

File: tests/core/test_lightning_optimizer.py (hunk 5 above)
Comment (tchaton) [reply tests/core/test_lightning_optimizer.py:586]: Not yet, in precision 16 with AMP native, when mocking step, we call scaler.step(optimizer). However, the step function being mocked, it doesn't get called internally and the count is 0.

File: pytorch_lightning/trainer/training_loop.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/trainer/training_loop.py]: I removed it. It appeared to me it was more a hack for debugging than anything else.

File: tests/trainer/optimization/test_parity_optimization.py [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+"""
+
+##################################################
+# TESTING AUTOMATIC OPTIMIZATION - ONE OPTIMIZER #
+##################################################
+
+################## MODULES ##################   # noqa E266
+
+
+class BaseParityAutomatiocOptimizationModel(BoringModel):
```
Comment (SeanNaren) [inline tests/trainer/optimization/test_parity_optimization.py:48]: Super clean I love it, just a small spelling mistake in the class names `Automatioc` -> `Automatic`

Comment (tchaton) [reply tests/trainer/optimization/test_parity_optimization.py:48]: Thanks !

**Comment by Borda** [comment]
why was it closed, does it have a followup?
